### ISSUE TITLE
Discussion: Where to relocate package tests?

### ISSUE THREAD

**Issue by aaraney** [NOAA-OWP/DMOD#229]
Currently, all `dmod` subpackages share a `test` subpackage. Meaning, all the tests from each `dmod` subpackage is collected and collated into a single directory. I am not intimately familiar with all of our testing suite, so I am likely wrong, but I would speculate this is so that common testing functionality can more easily be shared. 

"installed" (`site-packages`) subpackage layout is as follows:
```shell
dmod
├── access
├── ...
├── subsetservice
└── test
```

Consequently, this means that it is possible for test files to be overwritten by another subpackage during installation if they share the same name. This is because in our source tree, each subpackage looks similar to this:

```shell
dmod
├── access
│   ├── __init__.py
│   └── ...
└── test
    ├── __init__.py
    └── test_some_cool_thing.py
```

This is not super likely, however, this also means that the last installed subpackage's `test/__init__.py` file will overwrite any prior installed versions. One way, as a team, we've talked about getting around this, is to keep all `test/__init__.py` files empty. I think that is sane and an easy fix, however I tend to think we can do better.

Aside from the aforementioned name mangeling issue, it would be nice to more easily run individual test files or individual tests. Due to our frequent usage of relative imports in tests and python's import semantics, running individual tests is doable, but often requires some manual intervention (adjusting imports, removing `__init__.py`). Likewise, it would be nice if popular IDE test runner plugins "just worked" out-of-the-box.

So, as a segue, here are my wants (please add yours as well):

1. ability to run tests at the package, file, and test scopes
2. ability to use tests in edit mode
3. shareable and reusable testing suite
4. straight forward IDE support

With the above friction-points and wants, the first thing that I think we should discuss and decide is, do we move tests? and if so, where too? If the answer is yes, I think there are two straightforward and logical options. 1) for each subpackage, we move it's test files inside the subpackage as a new subsubpackage (i.e. `dmod.access.test`) or 2) we move the tests outside of installed source scope up one level so they sit next to a subpackage's `setup.py` and `README.md`. Both of these locations have a similar drawback. Without a testing framework or library, code reusable is a bit more challenging (back to that in a minute). However, the main difference is, do you ship tests with installed code or not? Given that this project is heavily application specific, I lean in the direction of shipping tests with the code. This in the future would allow us to run our test suite on some target whenever we would like. Now, given our use of containerization, I can already hear the pushback, but I think given python being the language, I could foresee running tests on a machine during a debugging procedure.

Once we've addressed the previous concern, I think the use of a testing framework like [`pytest`](https://docs.pytest.org/en/7.2.x/) could easily remedy the reusability want. `pytest` has an extensive plugin environment and makes it trivial to share your own plugins or [`fixtures`](https://docs.pytest.org/en/6.2.x/fixture.html) across a codebase. And yes, all of our current tests will just work. It's super easy to use and I think makes writing tests more enjoyable and a lot faster (check out [`pytest.parametrize`](https://docs.pytest.org/en/6.2.x/parametrize.html)). If it's not become clear, I am a huge fan and advocate of `pytest`. It's used internally in `hydrotools`, `hypy`, and `ngen-cal` and externally in codebases like `numpy`, `pandas`, and `sklearn`. It's super easy to pick up and I even have a transition guide written for people coming from python's `unittest` library.

**Comment by christophertubbs** [comment]
The overwriting scenario has already been encountered. If you're using an IDE like PyCharm, moving files within the test directory will create new `__init__.py` files as necessary in order to maintain package/subpackage structure. Have a bajillion unit tests for a library and want to reorganize? Moving those files from within the IDE will create a new `__init__.py` and override the `__init__.py` that's needed in order to link some tests to their helper classes. The easiest solution under this model is to further nest the tests, ending up with something like:
```
dmod
├── access
│   ├── __init__.py
│   └── ...
└── test
    └── access
        ├── __init__.py
        └── test_some_cool_thing.py
```

Works, but kinda ugly. Your relative imports there become ugly too because you're not doing something like `from ..access import whatever`, you're doing `from ...access import whatever` and it just escalates from there.

Some of this may be cleaned up by flattening the source tree. The current structure does _not_ match what gets install in your virtual environment, resulting in some weird bugs from time to time (like the overriding issue).

What I'd _like_ to see is something like:

```
DMOD
├── dmod
|    ├── __init__.py
|    ├── access
|    |   ├── __init__.py
|    |   └── ...
|    ├── requestservice
|    |   ├── __init__.py
|    |   └── ...
|    └── ...
|       ├── __init__.py
|       └── ...
├── test
|   ├── __init__.py
|   ├── access
|   |  ├── __init__.py
|   |  └── test_some_cool_thing.py
|   ├── requestservice
|   |   ├── __init__.py
|   |   └── ...
|   └── ...
|       ├── __init__.py
|       └── ...
├── setup.py
└── test_dmod.py
```

`test_dmod.py` can be the access point for tests (to ensure paths can work) and the working directory for nested tests can be set to the root of the project when running individual tests. Unless there's some hocus pocus I'm not aware of, you can't really do direct test runs at the moment since you'll run into fun errors due python not being able to climb up higher than the calling directory when running them, but directing the tests to work like that prevents them from working when running the tests on installed directories. Format it like this and you _should_ be able to avoid some pathing headaches. I think this might also render `./scripts/update_package.sh` unnecessary since you'll no longer need to install dependent subpackages in order to reference them by name.

Also, `setup.py` and such should probably ensure that the tests don't end up in the finished product. That's currently the case and that means that required testing input files end up in the virtual environment.

**Comment by robertbartel** [comment]
> Currently, all dmod subpackages share a test subpackage ... I would speculate this is so that common testing functionality can more easily be shared.

Not exactly.  The idea was to have identical structure for clarity, but the intent was also to _avoid_ installing the `dmod.test` namespace package.  That's the purpose of the `exclude` usage in the lines like this in most all (if not all) package _setup.py_ files (it's just that it was done incorrectly and never fixed):

```
packages=find_namespace_packages(exclude=('tests',` 'schemas', 'ssl', 'src'))
```

Putting common testing functionality somewhere does sound like a good idea, but it should all be in one place and somewhere else. 

> This is not super likely, however, this also means that the last installed subpackage's test/__init__.py file will overwrite any prior installed versions.

I don't think we really want to ever install tests; we just need to fix the bug(s) that does that.   Testing should occur before a package is installed.  This is why the tests modules for a given package use relative imports for things in that package:  they assume the things they are intended to test are not yet installed.

We can discuss whether tests should actually be installed more if needed; if so, I would recommend we settle that first.

**Comment by aaraney** [comment]
Thanks for the input, @christophertubbs! I am not going to paste your above file structure layout, but this is exactly what I was thinking as another option if we _don't_ want to move tests inside of subpackages. Thanks for making it more clear.

> Unless there's some hocus pocus I'm not aware of, you can't really do direct test runs at the moment since you'll run into fun errors due python not being able to climb up higher than the calling directory when running them, but directing the tests to work like that prevents them from working when running the tests on installed directories. Format it like this and you should be able to avoid some pathing headaches.

Yeah, you have to do a little hocus pocus. Ive been getting around this by changing relative imports in tests to absolute imports and removing `__init__.py` from a subpackage's `test/` \~subpackage\~. Then using `pytest` to run the tests (i.e. `pytest test_some_cool_feature.py`).

> I think this might also render ./scripts/update_package.sh unnecessary since you'll no longer need to install dependent subpackages in order to reference them by name.

I agree. I don't think we will need a separate script either, we should be able to run `pytest` from the root of the project and it will discover and run all the tests for us.

**Comment by robertbartel** [comment]
> Aside from the aforementioned name mangeling issue, it would be nice to more easily run individual test files or individual tests. 

The _scripts/test_package.sh_ script does have some capabilities for isolating the tests to run more specifically than just package/directory.  We can expand that further if there are useful scenarios it doesn't support yet.  I probably didn't add something like that already because I just ran the tests through Intellij in those situations, which segues into ...

> Due to our frequent usage of relative imports in tests and python's import semantics, running individual tests is doable, but often requires some manual intervention (adjusting imports, removing __init__.py). Likewise, it would be nice if popular IDE test runner plugins "just worked" out-of-the-box.

With perhaps 2 or 3 exceptions, I have not had trouble running tests as-is via Intellij, either by test class (that isn't exactly the same as by file, but I wonder if that's the use-case you are really interested in) or by individual test function.  I do remember a couple times when a test function encountered a strange error when run through the IDE, but passed just fine when run through _scripts/test_package.sh_.  

There is probably an underlying issue somewhere (or perhaps a few) that needs to be addressed and/or improved, but in principle, what you suggest is essentially already possible.  Other IDEs may behave differently, but I would expect it to be fairly easy to get them to work (now we're getting into exactly what you meant by "'just worked' out-of-the-box").   Regardless, I am curious what problems you have encountered that prompted this.

**Comment by aaraney** [comment]
> The scripts/test_package.sh script does have some capabilities for isolating the tests to run more specifically than just package/directory. 

> With perhaps 2 or 3 exceptions, I have not had trouble running tests as-is via Intellij, either by test class (that isn't exactly the same as by file, but I wonder if that's the use-case you are really interested in) or by individual test function.

I figured that was the case, I think the help docs might just a little out of date. I was wanting to run specific tests using either regex or the test name itself.

> There is probably an underlying issue somewhere (or perhaps a few) that needs to be addressed and/or improved, but in principle, what you suggest is essentially already possible. 

Right, I think vscode isn't liking that tests are in their own namespace, so when a test module goes to do a relative import from the subpackage, it freaks out.

**Comment by robertbartel** [comment]
> I don't think we will need a separate script either, we should be able to run pytest from the root of the project and it will discover and run all the tests for us.

Some of the need to adopt of testing helper scripts has probably come from not making use of capabilities otherwise available, but some has also been due to the complexity of the project.  

Consider, e.g., an integration test that requires the `object_store` Docker stack to be running (`IntegrationTestObjectStoreDatasetManager` ).  It doesn't need the stack started freshly for each individual test invocation:  just that it be running.  Assume that starting the stack takes 2 minutes on a development machine, and that multiple test invocations - perhaps in rapid succession - are going to be performed as changes to the tested code are made.

It introduces an interesting design question of whether (assuming it can) **pytest** _should_ handle the isolated setup (i.e., without necessarily running any test invocations) and eventual isolated teardown of these and other such things, versus a project-specific tool.  And then dominos start to fall on what else whatever is chosen should also be capable of handling.   

That's not necessarily an argument for keeping these scripts.  We just need to keep in mind there are some complex emergent properties of the project that may have subtle impacts or be subtly impacted by these things.

**Comment by aaraney** [comment]
> It introduces an interesting design question of whether (assuming it can) **pytest** _should_ handle the isolated setup (i.e., without necessarily running any test invocations) and eventual isolated teardown of these and other such things, versus a project-specific tool. And then dominos start to fall on what else whatever is chosen should also be capable of handling.
> 
> That's not necessarily an argument for keeping these scripts. We just need to keep in mind there are some complex emergent properties of the project that may have subtle impacts or be subtly impacted by these things.

Right, I agree and appreciate the brake pumping. There is always a balance between using a tool vs building and maintaining your own. I am not advocating to remove the `run_test.sh` script, but instead see how we could use `pytest` as a core test runner and framework for writing new tests that is compliant and interoperable with our existing testing scripts.

Aside from this, I think it would be helpful for us to first decide if tests should be moved elsewhere in the source tree? That decision should be agnostic of any candidate test runner.

**Comment by aaraney** [comment]
> The _scripts/test_package.sh_ script does have some capabilities for isolating the tests to run more specifically than just package/directory. We can expand that further if there are useful scenarios it doesn't support yet. I probably didn't add something like that already because I just ran the tests through Intellij in those situations, which segues into ...
> 
> > Due to our frequent usage of relative imports in tests and python's import semantics, running individual tests is doable, but often requires some manual intervention (adjusting imports, removing **init**.py). Likewise, it would be nice if popular IDE test runner plugins "just worked" out-of-the-box.
> 
> With perhaps 2 or 3 exceptions, I have not had trouble running tests as-is via Intellij, either by test class (that isn't exactly the same as by file, but I wonder if that's the use-case you are really interested in) or by individual test function. I do remember a couple times when a test function encountered a strange error when run through the IDE, but passed just fine when run through _scripts/test_package.sh_.
> 
> There is probably an underlying issue somewhere (or perhaps a few) that needs to be addressed and/or improved, but in principle, what you suggest is essentially already possible. Other IDEs may behave differently, but I would expect it to be fairly easy to get them to work (now we're getting into exactly what you meant by "'just worked' out-of-the-box"). Regardless, I am curious what problems you have encountered that prompted this.

To revisit this, I found my underlying issue (shocker @robertbartel was correct). I'm sorry for jumping to conclusions earlier. I should have done more investigating.

TLDR; `pytest` has an option I needed to specify -- `--import-mode=importlib`.

For background, I really like using `pytest` to run pretty much any python project's unit tests. It has just become part of my workflow and I think I am a little guilty of forcing `pytest` as a solution to a problem that doesn't exist. So, my apologies for that. In reading the `pytest` docs I discovered that `pytest` does not, by default, use the same module import strategy that `unittest` does. `unittest` uses `importlib` and `pytest` uses their own import strategy. Funny enough, `pytest` [now recommends](https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#choosing-a-test-layout-import-rules) using `importlib` over their own implementation. Im sure with `PEP`s over the years related to implicit namespace packages for example, these and other features have been added to `importlib` and `pytest` has just said, use `importlib` instead -- makes sense. All of this to say, all I needed to do was specify `--import-mode=importlib` to `pytest` and it all worked. You can also specify this is several of their supported configuration files. Once I specified it in a configuration file, my IDE happily ran tests and used `pytest` as the test runner.

Moving forward, I don't know the we _need_ to move the tests. Now knowing that tests were not intended to be packaged in the first place:

> ... The idea was to have identical structure for clarity, but the intent was also to _avoid_ installing the `dmod.test` namespace package. That's the purpose of the `exclude` usage in the lines like this in most all (if not all) package _setup.py_ files (it's just that it was done incorrectly and never fixed):
> 
> ```
> packages=find_namespace_packages(exclude=('tests',` 'schemas', 'ssl', 'src'))
> ```

It seems that once that is corrected, we will not have namespace mangling issues in `dmod.test.__init__.py`. @christophertubbs, @robertbartel is that fair, meaning we can close this? Or what other concerns have not been addressed?

**Comment by robertbartel** [comment]
> With perhaps 2 or 3 exceptions, I have not had trouble running tests as-is via Intellij, either by test class (that isn't exactly the same as by file, but I wonder if that's the use-case you are really interested in) or by individual test function.

Well, this used to be the case (and is again now).

I started to have tests run through Intellij fail to find the test class.  It turned out that Intellij had effectively changed the default Python test runner it was using (_Preferences->Tools->Python Integrated Tools->Testing_).  This was **Autodetect** instead of **UnitTest**.   It isn't clear whether this was changed directly or the automatic detection rules were changed; regardless, now the **Twisted Trial** runner was actually being used.  Once I changed things to make sure the **UnitTest** runner was used, I was able to run tests through the IDE again. 

@aaraney, I'm curious if any issues with tests in IDEs that you and others have seen are due to something similar.

**Comment by robertbartel** [comment]
@aaraney, I hadn't seen your most recent comment, but it lines up with what I just experienced.

I've opened #241 to address the that issue.  With that being tracked separately, yes, I think we can close this.  I suspect #207 can be closed also.  I'll give @christophertubbs a chance to chime in, though.

**Comment by aaraney** [comment]
> @aaraney, I'm curious if any issues with tests in IDEs that you and others have seen are due to something similar.

I use VSCode and I think it is an issue, similar to your case, with test discovery. I did a little digging and VSCode is relying on [`unittest.TestLoader.discover`](https://docs.python.org/3/library/unittest.html#unittest.TestLoader.discover) to discover tests. This is analogous to using `python -m unittest discover`. For what it is worth, I cannot get `unittest discover` to discover and run the tests. Ive tried from both the root level of the repo, `./python/lib`, and `./python/lib/{{ package }}`. Ive even tried removing the `gui` directory completely as I was running into `django` issues. So I am scratching my head a bit.

I was, however, able to get `pytest` to working with a few workarounds. 

1. Added `conftest.py` to repo root, this is a `pytest` configuration file. In short, `pytest` needs to append the repo root to `sys.path`.

```python
# conftest.py
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).parent.resolve()

# this is analogous to running `python -m pytest`
sys.path.insert(0, str(REPO_ROOT))
```

2. Created a `pytest.ini` file at the repo root.

```ini
# pytest.ini
[pytest]
addopts = --import-mode=importlib
```

3. Configured Python testing in VSCode (`cmd (ctrl) + shift + p` -> `Python: Configure Tests` -> `pytest`)

4. Opened workspace setting in VSCode (`cmd (ctrl) + shift + p` -> `Preferences: Open Workspace Setting (JSON)`) and added the following:

```json
{
    "python.testing.pytestEnabled": true,
    "python.testing.pytestArgs": [
        "python/lib"
    ]
}
```

After that, testing and test discovery is working for me.

**Comment by christophertubbs** [comment]
> I've opened https://github.com/NOAA-OWP/DMOD/issues/241 to address the failure to exclude test sources as intended. With that being tracked separately, yes, I think we can close this. I suspect https://github.com/NOAA-OWP/DMOD/issues/207 can be closed also. I'll give @christophertubbs a chance to chime in, though.

I think the majority of my issues were just with testing Django stuff under `python/services/evaluationservice`, which should probably be tested through `manage.py` anyways, so everything sounds good to me. Excluding the test directories from the `setup.py` files will probably fix everything else I ran into (can't remember which). I want to say tests were getting mixed up and referencing test resources in the virtual environment rather than what was local. Exclude those test files and that scenario shouldn't crop up.

**Comment by robertbartel** [comment]
@aaraney, item 1 and 2 seem generic enough that we could add those to the repo.  It may be better for the moment, though, to document this (similarly to what I go into below) and add entries to _.gitignore_.  Potentially, we could also provide example files in some directory that people had to move into place themselves if they wanted. 

Item 3 and 4 are certainly something user-specific, though there is likely to be a wide spread need for that info.  The top-level _TESTING.md_ file has details on testing setup and tools, and a section on how to avoid our recent VSCode and Intellij issues might be helpful.

If/when we fully move to **pytest**, _TESTING.md_  will need to be updated, as it details our use of **unittest**.  That's another reason why I suggested the alternative of documenting+ignoring your **pytest** files:  we would have to review and updated this file (and _at least_ test the functionality of the testing scripts) if we do a full, official switch.

Thoughts?

**Comment by aaraney** [comment]
> @aaraney, item 1 and 2 seem generic enough that we could add those to the repo. It may be better for the moment, though, to document this (similarly to what I go into below) and add entries to _.gitignore_. Potentially, we could also provide example files in some directory that people had to move into place themselves if they wanted.

I agree that these are generic enough to merit moving them into the repo. Options specified in 2. can be specified in [`pyproject.toml`](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/)(pre-build requirements file; a file we _should_ add in the future next to package's `setup.py` / `setup.cfg`) as an alternative and maybe more attractive solution. 

> Item 3 and 4 are certainly something user-specific, though there is likely to be a wide spread need for that info. The top-level _TESTING.md_ file has details on testing setup and tools, and a section on how to avoid our recent VSCode and Intellij issues might be helpful.

Yeah, moving these into the testing documentation seems like the most sane option. There is likely some `pytest` configuration that I am unaware of that might make 4. unnecessary. 

> If/when we fully move to **pytest**, _TESTING.md_ will need to be updated, as it details our use of **unittest**. That's another reason why I suggested the alternative of documenting+ignoring your **pytest** files: we would have to review and updated this file (and _at least_ test the functionality of the testing scripts) if we do a full, official switch.

Right, this isn't the most pressing matter and, like you said, it will require a negligible amount of work, so I agree that punting this to a later discussion is probably the best idea. I foresee reevaluating this before we port services to REST a good timeline IMO. I think if we do switch, we will feel to see the benefits of using `pytest` when more complex testing fixtures and test parametrization combinations are required.

With that said, in the short term, I think we should add the functionality of 1. and 2 to the repo and leave 3. and 4. out. How is that compromise?

**Comment by robertbartel** [comment]
> With that said, in the short term, I think we should add the functionality of 1. and 2 to the repo and leave 3. and 4. out. How is that compromise?

I may be misunderstanding.  I don't think we should add those **pytest** files without also updating the testing documentation to explain them, at which point we should outline the VSCode test issues and how the user can apply steps 3 and 4 to address them.  We (i.e., I) probably also need to document the recent test issues in Intellij.  

We don't need to add or update VSCode workspace or settings files, though, assuming those typically exist locally inside the repo directory structure.

**Comment by aaraney** [comment]
Leaving a note for myself and anyone else who wants to use `pytest` with dmod. Here are the two files that i've added to my local dev environment to make `pytest` happy. This includes the ability to run `dmod` integration tests using an added custom `pytest` option, `--it`. You Put both of theses files in the root of your `dmod` clone / fork.

`pytest.ini`
[pytest]
addopts = --import-mode=importlib
; key=value pairs with no spaces
it_env_vars =
    ; A name for the Docker container used to run a Redis instance during integration testing
    IT_REDIS_CONTAINER_NAME="it_redis_container"
    ; The port-forwarded port on the host machine for the containerized Redis service
    IT_REDIS_CONTAINER_HOST_PORT=19639
    ;The containerized Redis service's password
    IT_REDIS_CONTAINER_PASS="DPXzqRqjhsXokOVQcPUqOJuzKePMsfUc"
    DOCKER_MPI_NET_NAME=mpi-net
    DOCKER_MPI_NET_SUBNET=10.0.0.0/24
    DOCKER_MPI_NET_GATEWAY=10.0.0.1
    DOCKER_MPI_NET_VXLAN_ID=4097
    DOCKER_REQUESTS_NET_NAME=requests-net
    DOCKER_REQUESTS_NET_SUBNET=10.0.1.0/27
    DOCKER_REQUESTS_NET_GATEWAY=10.0.1.1
    TEST_SSL_CERT_DIR=<your-abs-path-to-ssl-certs>
    PROJECT_ROOT=<your-abs-path-to-proj-root>
    MODEL_EXEC_ACCESS_KEY=dmod
    MODEL_EXEC_SECRET_KEY=password
```

`conftest.py`
```
import os
import sys
from pathlib import Path
import logging
import pytest
from typing import Dict, List, Set, TYPE_CHECKING, Optional
from pytest import Item, CallInfo, Session
import subprocess

logger = logging.getLogger("dmod_conftest")

if TYPE_CHECKING:
    from pytest import Config, Parser

REPO_ROOT = Path(__file__).parent.resolve()

# this is analogous to running `python -m pytest`
# this is needed for `pytest` discovery reasons specific to `dmod`'s package structure
sys.path.insert(0, str(REPO_ROOT))

integration_testing: bool = False
modules: Set[Path] = set()
active: Set[Path] = set()


def integration_testing_flag():
    global integration_testing
    integration_testing = True
    logger.debug("integration testing flag on")


def pytest_addoption(parser: "Parser"):
    parser.addoption(
        "--it", action="store_true", default=False, help="run integration tests"
    )
    parser.addini(
        "it_env_vars", "Environment variables for integration tests", type="args"
    )


def parse_it_env_vars(name_values: List[str]) -> Dict[str, str]:
    return dict(map(lambda pair: pair.split("="), name_values))


def pytest_configure(config: "Config"):
    if config.getoption("it"):
        python_files = config.getini("python_files")
        assert isinstance(python_files, list)
        python_files[:] = ["it_*.py"]

        it_env_vars = config.getini("it_env_vars")
        integration_testing_flag()

        parsed_vars = parse_it_env_vars(it_env_vars)
        logger.debug(f"Adding these environment variables: {parsed_vars}")

        os.environ.update(parsed_vars)


def get_setup_script_path(module: Path) -> Path:
    return module / "setup_it_env.sh"


def setup(setup_script: Path) -> "subprocess.CompletedProcess[bytes]":
    cmd = " ".join(["source", str(setup_script), "&&", "do_setup"])
    logger.debug(f"launching setup script: {cmd!r}")
    return subprocess.run(
        cmd,
        capture_output=True,
        shell=True,
    )


def teardown(setup_script: Path) -> "subprocess.CompletedProcess[bytes]":
    cmd = " ".join(["source", str(setup_script), "&&", "do_teardown"])
    logger.debug(f"launching teardown script: {cmd!r}")
    return subprocess.run(
        cmd,
        capture_output=True,
        shell=True,
    )


def format_completed_process(cp: "subprocess.CompletedProcess[bytes]") -> str:
    return (
        f"args:   {cp.args!r}\n"
        f"stdout: {cp.stdout.decode('utf-8').strip()}\n"
        f"stderr: {cp.stderr.decode('utf-8').strip()}"
    )


def pytest_runtest_setup(item: Item):
    if not integration_testing:
        return

    module = item.path.parent
    if module not in modules:
        modules.add(module)

        setup_script = get_setup_script_path(module)

        if setup_script.exists():
            out = setup(setup_script)
            logger.info(f"setup: {module.parent.parent.name!r}")
            logger.debug(format_completed_process(out))

            if out.returncode != 0:
                logger.debug(
                    f"non-zero return code ({out.returncode:3}): {module.parent.parent.name!r}"
                )
                logger.debug(format_completed_process(out))

                out = teardown(setup_script)
                logger.info(f"tearing down {module.parent.parent.name!r} early")

                if out.returncode != 0:
                    logger.debug(
                        f"non-zero return code ({out.returncode:3}): {module.parent.parent.name!r}"
                    )
                    logger.debug(format_completed_process(out))

                item.session.shouldstop = True
                # dont add to active
                return

            active.add(module)


def pytest_runtest_teardown(item: Item, nextitem: Optional[Item]):
    if not integration_testing:
        return

    next_item_module = nextitem.path.parent if nextitem is not None else None
    if next_item_module is None or next_item_module not in modules:
        # do item tear down
        module = item.path.parent
        setup_script = get_setup_script_path(module)

        out = teardown(setup_script)
        logger.info(f"tearing down: {module.parent.parent.name!r}")
        logger.debug(format_completed_process(out))

        active.remove(module)


def pytest_sessionfinish(session: Session, exitstatus: int):
    if not integration_testing:
        return

    while True:
        try:
            module = active.pop()
        except KeyError:
            break

        setup_script = get_setup_script_path(module)
        out = teardown(setup_script)

        logging.info(f"tearing down: {module.parent.parent.name!r}")
        logging.debug(format_completed_process(out))

```
